test/python/pulse/test_builder.py

```diff
@@ -0,0 +1,969 @@
+
+    def test_call_circuit_with_cregs(self):
+        """Test calling of circuit wiht classical registers."""
+
+        qc = circuit.QuantumCircuit(2, 2)
+        qc.h(0)
+        qc.cx(0, 1)
+        qc.measure([0, 1], [0, 1])
+
+        with pulse.build(self.backend) as schedule:
+            pulse.call(qc)
+
+        reference_qc = compiler.transpile(qc, self.backend)
+        reference = compiler.schedule(reference_qc, self.backend)
+
+        self.assertEqual(schedule, reference)
+
+
+class TestBuilderComposition(TestBuilder):
+    """Test more sophisticated composite builder examples."""
+
+    def test_complex_build(self):
+        """Test a general program build with nested contexts,
+        circuits and macros."""
+        d0 = pulse.DriveChannel(0)
+        d1 = pulse.DriveChannel(1)
+        d2 = pulse.DriveChannel(2)
+        delay_dur = 19
+        short_dur = 31
+        long_dur = 101
+
+        with pulse.build(self.backend) as schedule:
+            with pulse.align_sequential():
+                pulse.delay(delay_dur, d0)
+                pulse.u2(0, pi/2, 1)
+            with pulse.align_right():
+                pulse.play(pulse_lib.Constant(short_dur, 0.1), d1)
+                pulse.play(pulse_lib.Constant(long_dur, 0.1), d2)
+                pulse.u2(0, pi/2, 1)
+            with pulse.align_left():
+                pulse.u2(0, pi/2, 0)
+                pulse.u2(0, pi/2, 1)
+                pulse.u2(0, pi/2, 0)
+            pulse.measure(0)
+
+        single_u2_qc = circuit.QuantumCircuit(2)
+        single_u2_qc.u2(0, pi/2, 1)
+        single_u2_qc = compiler.transpile(single_u2_qc, self.backend)
+        single_u2_sched = compiler.schedule(single_u2_qc, self.backend)
+
+        triple_u2_qc = circuit.QuantumCircuit(2)
+        triple_u2_qc.u2(0, pi/2, 0)
+        triple_u2_qc.u2(0, pi/2, 1)
+        triple_u2_qc.u2(0, pi/2, 0)
+        triple_u2_qc = compiler.transpile(triple_u2_qc, self.backend)
+        triple_u2_sched = compiler.schedule(triple_u2_qc,
+                                            self.backend,
+                                            method='alap')
+
+        # outer sequential context
+        outer_reference = pulse.Schedule()
+        outer_reference += instructions.Delay(delay_dur, d0)
+        outer_reference.insert(delay_dur, single_u2_sched, mutate=True)
+
+        # inner align right
+        align_right_reference = pulse.Schedule()
+        align_right_reference += pulse.Play(
+            pulse_lib.Constant(long_dur, 0.1), d2)
+        align_right_reference.insert(long_dur-single_u2_sched.duration,
+                                     single_u2_sched,
+                                     mutate=True)
+        align_right_reference.insert(
+            long_dur-single_u2_sched.duration-short_dur,
+            pulse.Play(pulse_lib.Constant(short_dur, 0.1), d1),
+            mutate=True)
+        # inner align left
+        align_left_reference = triple_u2_sched
```

Comment:
This doesn't help me verify the input. It forces me to track an alias for a variable name so that when I get to line 903, I need to start re-reading the function from the start to understand what is going on. 862-869 should be moved here with the name `align_left_reference`

Reply:
Fixed in 0752e2140